tests/importer/importer.cpp

```diff
@@ -0,0 +1,427 @@
+    // Passing in one of the models for resolution will load modelD into the library.
+    auto importer = libcellml::Importer::create();
+    importer->resolveImports(modelA, resourcePath("resolveimports/"));
+
+    EXPECT_EQ(size_t(1), importer->libraryCount());
+
+    // Now resolve the other models and expect the library contents to be unchanged.
+    importer->resolveImports(modelB, resourcePath("resolveimports/"));
+    importer->resolveImports(modelC, resourcePath("resolveimports/"));
+
+    EXPECT_EQ(size_t(1), importer->libraryCount());
+}
+
+TEST(Importer, addModelToLibrary)
+{
+    // This test shows how a model instance can be manually added to the import library by the user,
+    // and will be used to resolve imports, rather than from an external file/url.
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("resolveimports/generic_no_source.cellml"));
+    auto importer = libcellml::Importer::create();
+
+    auto sourceModel = libcellml::Model::create("source");
+    sourceModel->addComponent(libcellml::Component::create("a"));
+    sourceModel->addComponent(libcellml::Component::create("b"));
+    sourceModel->addComponent(libcellml::Component::create("c"));
+    sourceModel->addComponent(libcellml::Component::create("d"));
+
+    // Add a model manually to the library, including the URL that it will replace in future imports.
+    EXPECT_TRUE(importer->addModel(sourceModel, "i_dont_exist.cellml"));
+
+    // Can't add to the same url key more than once.
+    EXPECT_FALSE(importer->addModel(sourceModel, "i_dont_exist.cellml"));
+
+    importer->resolveImports(model, "");
+
+    EXPECT_EQ(size_t(0), importer->issueCount());
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(Importer, replaceModel)
+{
+    // This test shows how a model instance can be manually added to the import library by the user,
+    // and will be used to resolve imports, rather than from an external file/url.
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("resolveimports/generic_no_source.cellml"));
+    auto importer = libcellml::Importer::create();
+
+    auto wrongSourceModel = libcellml::Model::create("wrong");
+
+    auto rightSourceModel = libcellml::Model::create("source");
+    rightSourceModel->addComponent(libcellml::Component::create("a"));
+    rightSourceModel->addComponent(libcellml::Component::create("b"));
+    rightSourceModel->addComponent(libcellml::Component::create("c"));
+    rightSourceModel->addComponent(libcellml::Component::create("d"));
+
+    // Add a model manually to the library, including the URL that it will replace in future imports.
+    EXPECT_TRUE(importer->addModel(wrongSourceModel, "i_dont_exist.cellml"));
+    // Can't add to the same url key more than once.
+    EXPECT_FALSE(importer->addModel(rightSourceModel, "i_dont_exist.cellml"));
+
+    // Use replaceModel instead.
```

Comment:
I know I should know this, but is this how our add methods work in other places?

Reply:
At the moment it's very inconsistent ... which is what https://github.com/cellml/libcellml/pull/621 will fix (nudge nudge) ... and after that's in, then yes, this will be in line with elsewhere.